webui/react/src/hooks/useCustomizeColumnsModal.tsx

```diff
@@ -0,0 +1,234 @@
+import { Button, Input, Modal, ModalFuncProps } from 'antd';
+import { ModalFunc } from 'antd/es/modal/confirm';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { FixedSizeList as List } from 'react-window';
+
+import { isEqual } from 'utils/data';
+import { camelCaseToSentence } from 'utils/string';
+
+import css from './useCustomizeColumnsModal.module.scss';
+
+interface ModalState {
+  columns: string[];
+  defaultVisibleColumns: string[];
+  initialVisibleColumns: string[];
+  onSave?: (columns: string[]) => void;
+  visible: boolean;
+}
+
+export interface ShowModalProps {
+  columns: string[];
+  defaultVisibleColumns: string[];
+  initialVisibleColumns: string[];
+  onSave?: (columns: string[]) => void;
+}
+
+interface ModalHooks {
+  showModal: (props: ShowModalProps) => void;
+}
+
+const useCustomizeColumnsModal = (): ModalHooks => {
+  const modalRef = useRef<ReturnType<ModalFunc>>();
+  const [ modalState, setModalState ] = useState<ModalState>({
+    columns: [],
+    defaultVisibleColumns: [],
+    initialVisibleColumns: [],
+    visible: false,
+  });
+  const [ searchTerm, setSearchTerm ] = useState('');
+  const [ visibleColumns, setVisibleColumns ] = useState<string[]>([]);
+
+  const showModal = useCallback((
+    { columns, initialVisibleColumns, defaultVisibleColumns, onSave }: ShowModalProps,
+  ) => {
+    setModalState({ columns, defaultVisibleColumns, initialVisibleColumns, onSave, visible: true });
+    setVisibleColumns(initialVisibleColumns);
+  }, []);
+
+  const closeModal = useCallback(() => {
+    if (!modalRef.current) return;
+    modalRef.current.destroy();
+    modalRef.current = undefined;
+  }, []);
+
+  const handleCancel = useCallback(() => {
+    if (!modalRef.current) return;
+    closeModal();
+  }, [ closeModal ]);
+
+  const handleSave = useCallback((state: ModalState) => {
+    if (!modalRef.current) return;
+    state.onSave?.(visibleColumns);
+    closeModal();
+  }, [ closeModal, visibleColumns ]);
+
+  const handleSearch = useCallback((e) => {
+    setSearchTerm(e.target.value.toLowerCase());
```

Comment:
**nit:** looking at how the search term is used, it seems it's used via regex pretty consistently with ignore case, so the `toLowerCase()` is probably unnecessary. I was initially going to suggest this to avoid mutating the search term of what the user types (ie. as the user type, changing their input on the fly to lower case) but that doesn't seem to be the case (the case is preserved) so it's not a big deal either way.